designs/RestoreAndBuildCompatibility.md

```diff
@@ -0,0 +1,179 @@
+
+* nuget.exe restore - which downloads a defined list of packages to a specific folder.
+* build - with your tooling of choice. Said tooling doesn't need to be aware of the fact that NuGet is involved in any way.
+
+### NuGet PackageReference
+
+Since the introduction of project.json and PackageReference (2015 & 2017 respectively), NuGet has become tightly integrated with the build.
+Today NuGet's PackageReference is the tooling of choice for *all* .NET Core, now one unified .NET, projects.
+
+With PackageReference projects a restore is now *a must*.
+In addition to `nuget.exe`, you can restore PackageReference based projects with `msbuild /t:restore` and `dotnet.exe restore`.
+
+The `restore` and `build` operations have a contract expressed through 3 files:
+
+* `project.assets.json`
+* `projectName.ext.nuget.g.props`
+* `projectName.ext.nuget.g.targets`
+
+At build-time these files are loaded and interpreted as necessary.
+This now meant that the NuGet version and build tooling version have an implied compatibility matrix.
+
+### CI experience and compatibility matrices 
+
+The additional tooling options, the combination of project types in one repository now means that there are many ways people run the restore/build gestures on CI.
+
+1. dotnet.exe
+With dotnet.exe, the CI commands would normally look like one of the below options: 
+
+    * dotnet.exe restore & dotnet.exe build --no-restore
+    * dotnet.exe build (restore is implied)
+
+This is the recommended and full proof way if you have only projects that are supported by the [`.NET SDK`](https://github.com/dotnet/sdk).
+packages.config projects *are not* supported with dotnet.exe.
+
+dotnet.exe and it's respective SDK are a self-contained build tool. The restore and build contract is 100% understood and there are no compatibility concerns.
+
+2. MSBuild.exe
+
+With MSBuild.exe, the CI commands would normally look like one of the below options:
+
+* MSBuild.exe /t:restore & MSBuild.exe /t:build
+* MSBuild.exe /restore /t:build
+
+MSBuild.exe has had PackageReference restore support since VS 2017. Starting with VS 2019, update 5, MSBuild now has project specific packages.config support.
+
+Using MSBuild.exe to restore and build the repo means that the restore and build contract is 100% understood.
+
+3. NuGet.exe
+
+NuGet.exe is the oldest tool that can be used to restore on CIs.
+It supports PackageReference and packages.config, but obviously it cannot build, so one still needs to build, usually using MSBuild.exe.
+So the steps are usually:
+
+* NuGet.exe restore
+* MSBuild.exe /t:build
+
+In packages.config, this usually just works because of packages.config's architecture.
+In PackageReference, there's a nuance, because now you have a restore and build steps that are from potential different versions of the tooling.  
+
+Naturally this leads to 2 potential different problems.
+
+#### Old NuGet tooling, newer build tooling - NuGet forward compatibility

```

Comment:
I think of this more as "build tooling backward compatibility" rather than "NuGet forward compat". It's not a promise that the outputs won't change, but a promise that new tooling will be able to understand older outputs.

Reply:
I get that :) Given that backward/forward compat is relative to the tooling, I'm remove the mentions and keep it as simple as `older` vs `newer`